src/text.c

```diff
@@ -0,0 +1,441 @@
+	// write and escape value
+	const char *p = v;
+	const char *n = p;
+	while (*n) {
+		switch (*n) {
+		case '\\':
+			SKIP;
+			write (fd, "\\\\", 2);
+			break;
+		case '\n':
+			SKIP;
+			write (fd, "\\n", 2);
+			break;
+		case '\r':
+			SKIP;
+			write (fd, "\\r", 2);
+			break;
+		}
+		n++;
+	}
+	FLUSH;
+}
+
+static bool text_save(Sdb *s, int fd, bool sort, SdbList *path) {
+	//path
+	write_path (fd, path);
+	write (fd, "\n", 1);
+
+	// k=v entries
+	SdbList *l = sdb_foreach_list (s, sort);
+	SdbKv *kv;
+	SdbListIter *it;
+	ls_foreach (l, it, kv) {
+		const char *k = sdbkv_key (kv);
+		const char *v = sdbkv_value (kv);
+
+		write_key (fd, k);
+		write (fd, "=", 1);
+		write_value (fd, v);
+
+		write (fd, "\n", 1);
+	}
+	ls_free (l);
+
+	// sub-namespaces
+	l = s->ns;
+	if (sort) {
+		l = ls_clone (l);
+		ls_sort (l, cmp_ns);
+	}
+	SdbNs *ns;
+	ls_foreach (l, it, ns) {
+		write (fd, "\n", 1);
+		ls_push (path, ns->name);
+		text_save (ns->sdb, fd, sort, path);
+		ls_pop (path);
+	}
+	if (l != s->ns) {
+		ls_free (l);
+	}
+
+	return true;
+}
+#undef FLUSH
+#undef SKIP
+
+SDB_API bool sdb_text_save_fd(Sdb *s, int fd, bool sort) {
+	SdbList *path = ls_new ();
+	if (!path) {
+		return false;
+	}
+	bool r = text_save (s, fd, sort, path);
+	ls_free (path);
+	return r;
+}
+
+SDB_API bool sdb_text_save(Sdb *s, const char *file, bool sort) {
+	int fd = open (file, O_WRONLY | O_CREAT | O_TRUNC);
+	if (fd < 0) {
+		return false;
+	}
+	bool r = sdb_text_save_fd (s, fd, sort);
+	close (fd);
+	return r;
+}
+
+typedef enum {
+	STATE_NEWLINE,
+	STATE_PATH,
+	STATE_KEY,
+	STATE_VALUE
+} LoadState;
+
+typedef struct {
+	int fd;
+	bool eof;
+	size_t bufsz;
+	char *buf;
+	size_t buf_filled; // how many bytes of buf are currently filled with data from the file
+	Sdb *root_db;
+	Sdb *cur_db; // current namespace, changes when encountering a path line
+	// can't use pointers here because buf may be moved on realloc
+	size_t pos; // current processing position in the buffer
+	size_t token_begin; // beginning of the currently processed token in the buffer
+	size_t shift; // amount to shift chars to the left (from unescaping)
+	SdbList/*<size_t>*/ *path;
+	LoadState state;
+	bool unescape; // whether the prev char was a backslash, i.e. the current one is escaped
+} LoadCtx;
+
+
+// load more data from ctx->f into ctx->buf starting at ctx->pos
+// and update ctx->buf_filled and ctx->eof.
+// may realloc and change ctx->buf and ctx->bufsz.
+static bool load_read_more(LoadCtx *ctx) {
+	// realloc if no space
+	if (ctx->pos >= ctx->bufsz - 1) {
+		size_t newsz = ctx->bufsz + ctx->bufsz;
+		if (newsz < ctx->bufsz || newsz > (size_t)INT_MAX) {
+			return false;
+		}
+		char *newbuf = realloc (ctx->buf, newsz);
+		if (!newbuf) {
+			return false;
+		}
+		ctx->buf = newbuf;
+		ctx->bufsz = newsz;
+	}
+	// read full buffer (leave 1 for null-terminating)
+	int red = read (ctx->fd, ctx->buf + ctx->pos, ctx->bufsz - 1 - ctx->pos);
```

Comment:
using slurp() you dont need to handle all the buffer caching by yourself. so i would vote for simpler implementation by just reading the whole file in memory (or plain mmap), and make the text_load() function take the filename as argument. itt will be faster and simpler

Reply:
I would like to avoid slurp since these text files can become quite large. But I will try to do it with mmap+cow, which might even work very similar to cdb and then I can embed it in the Sdb struct if possible.